drakrun/drakrun/storage.py

```diff
@@ -308,9 +310,203 @@ def import_vm0(self, path: str):
+            )
+        except subprocess.CalledProcessError:
+            raise RuntimeError("Failed to create a new volume using lvcreate.")
+
+    def snapshot_vm0_volume(self):
+        """ Saves or snapshots base vm-0 volume for later use by other VMs """
+        # vm-0 is the original disk being treated as a snapshot
+        # vm-0-snap is being created just for the access time of the change in vm snapshot
+        try:
+            subprocess.check_output(
+                [
+                    "lvcreate",
+                    "-s",
+                    "-L", self.snapshot_disksize,
+                    "-n", "vm-0-snap",
+                    f"{self.lvm_volume_group}/vm-0"
+                ],
+                stderr=subprocess.STDOUT
+            )
+        except subprocess.CalledProcessError as exc:
+            logging.debug(exc.output)
+            raise RuntimeError("Couldn't create snapshot")
+
+    def get_vm_disk_path(self, vm_id: int) -> str:
+        """ Returns disk path for given VM as defined by XL configuration """
+        return f"phy:/dev/{self.lvm_volume_group}/vm-{vm_id},hda,w"
+
+    def rollback_vm_storage(self, vm_id: int):
+        """ Rolls back changes and prepares fresh storage for new run of this VM """
+        vm_id_vol = os.path.join("/dev", f"{self.lvm_volume_group}", f"vm-{vm_id}")
+
+        if vm_id == 0:
+            raise Exception("vm-0 should not be rollbacked")
+
+        logging.info(f"Rolling back changes to vm-{vm_id} disk")
+        if os.path.exists(vm_id_vol):
+            try:
+                subprocess.check_output(
+                    [
+                        "lvremove",
+                        "-v",
+                        "-y",
+                        # "--noudevsync",
+                        f"{self.lvm_volume_group}/vm-{vm_id}"
+                    ],
+                    stderr=subprocess.STDOUT,
+                )
+            except subprocess.CalledProcessError as exc:
+                logging.debug(exc.output)
+                raise RuntimeError(f"Failed to discard previous logical volume {self.lvm_volume_group}/vm-{vm_id}")
+
+        try:
+            subprocess.check_output(
+                [
+                    "lvcreate",
+                    "-s",
+                    "-L", self.snapshot_disksize,
+                    "-n", f"vm-{vm_id}",
+                    f"{self.lvm_volume_group}/vm-0"
+                ],
+                stderr=subprocess.STDOUT
+            )
+        except subprocess.CalledProcessError as exc:
+            logging.debug(exc.output)
+            raise RuntimeError("Couldn't rollback disk")
+
+    def get_vm0_snapshot_time(self):
+        """ Get UNIX timestamp of when vm-0 snapshot was last modified """
+
+        dir_path = '/etc/lvm/archive'
+        proc = subprocess.run(f"ls {dir_path} -t | grep {self.lvm_volume_group} | head -n 1", shell=True, stdout=subprocess.PIPE)
+        file_name = proc.stdout.decode().strip()
```

Comment:
Sure, will update that. Any specific reason for not using bash one-liner though?

Reply:
> Sure, will update that. Any specific reason for not using bash one-liner though?

Ah, got the reason myself only. Not everyone might be using bash so it might break